l3kernel/l3bitset.dtx

```diff
@@ -0,0 +1,586 @@
+              {
+                \str_put_left:Nn #1 { 0 }
+              }
+            \str_put_left:Nn #1 { 1 }
+          }
+          {
+            %replace value
+            \str_set:Nx #1
+              {
+                \str_range:Nnn #1 {1}{-1 - (#2)}
+                1
+                \str_range:Nnn #1 {1 -(#2)}{-1}
+             }
+          }
+      }
+  }
+
+%#1 name, #2 index (integer expression,1-based)
+\cs_new_protected:Npn \bitset_gset_true:Nn #1 #2
+  {
+    \int_compare:nNnT {#2 } > { 0 }
+      {
+        \int_compare:nNnTF {\str_count:N #1 } < { #2  }
+          {
+            %extend the str
+            \int_step_inline:nn { #2 - \str_count:N #1 -1}
+              {
+                \str_gput_left:Nn #1 { 0 }
+              }
+            \str_gput_left:Nn #1 { 1 }
+          }
+          {
+            % replace the value
+            \str_gset:Nx #1
+              {
+                \str_range:Nnn #1 {1}{-1 -(#2)}
+                1
+                \str_range:Nnn #1 {1 -(#2)}{-1}
+              }
+          }
+      }
+  }
+\cs_generate_variant:Nn \bitset_set_true:Nn  {cn}
+\cs_generate_variant:Nn \bitset_gset_true:Nn {cn}
+%    \end{macrocode}
+% \end{macro}
+
+% \begin{macro}
+%   {
+%     \bitset_set_false:Nn,  \bitset_set_false:cn,
+%     \bitset_gset_false:Nn, \bitset_gset_false:cn
+%   }
+% Unsetting a bit has only to do something if the string is longer than then index.
+%    \begin{macrocode}
+\cs_new_protected:Npn \bitset_set_false:Nn #1 #2  %#1 name, #2 index (1-based)
+ {
+   \int_compare:nNnT {#2 } > { 0 }
+     {
+       \int_compare:nNnT {\str_count:N #1 } > { #2 -1 }
+         {
+           % need to replace the str
+           \str_set:Nx #1
+             {
+               \str_range:Nnn #1 {1}{-1 - (#2)}
+                0
+               \str_range:Nnn #1 { 1 - (#2)}{-1}
+             }
+         }
+     }
+ }
+
+\cs_new_protected:Npn \bitset_gset_false:Nn #1 #2  %#1 name, #2 index (1-based)
+ {
+   \int_compare:nNnT {#2 } > { 0 }
+     {
+       \int_compare:nNnT {\str_count:N #1 } > { #2 -1 }
+         {
+           % need to replace the str
+           \str_gset:Nx #1
+             {
+               \str_range:Nnn #1 {1}{-1 -(#2)}
+                0
+               \str_range:Nnn #1 {1 - (#2)}{-1}
+             }
+         }
+     }
+ }
+\cs_generate_variant:Nn \bitset_set_false:Nn {cn}
+\cs_generate_variant:Nn \bitset_gset_false:Nn {cn}
+%    \end{macrocode}
+%  \end{macro}
+% \begin{macro}
+%   {
+%     \bitset_to_arabic:N, \bitset_to_arabic:c,
+%     \bitset_to_bin:N,    \bitset_to_bin:c,
+%   }
+%   The naming of the commands follow the names in the \texttt{int} module.
+%   \cs{bitset_to_arabic:N} uses \cs{int_from_bin:n} and so can only handle small
+%   numbers. There is no test for the string length, so that long bitset with
+%   many leading zeros work nevertheless.
+%    \begin{macrocode}
+\cs_new:Npn \bitset_to_arabic:N #1
+  {
+    \exp_args:Ne \int_from_bin:n { #1 }
+  }
```

Comment:
Should we have a 'size' check here, with at present an error of the form 'Please discuss this with the team' or similar. We likely need to integrate `bigint` or at least a subset; perhaps Bruno can comment on that aspect. (I think division was the outstanding function.)

Reply:
It's been a while.  Yes, division was still outstanding, but more importantly the code could be drastically simplified now that we assume `\expanded` is available.  I had some pretty elaborate but now useless code to convert big integers to some kind of tree to speed up all the operations, but it makes the logic extremely complicated, which stopped me from implementing division.

Reply:
I wonder if we can/should use fp for now for  larger bitsets. It isn't superfast for large bitsets (but probably the speeding-up-experts can do something here), but at least it doesn't error:

~~~~
\documentclass{article}
\begin{document}
\ExplSyntaxOn
\bitset_new:N \l_my_bitset
\bitset_set_true:Nn \l_my_bitset {1}
\bitset_set_true:Nn \l_my_bitset {25}
\bitset_set_true:Nn \l_my_bitset {14}

\cs_new:Npn \__bitset_bin_factor:n #1 { + \str_item:Nn\l_my_bitset{-#1} * 2**(#1-1) }

\cs_set:Npn \bitset_to_arabic:N #1
  {
    \int_compare:nNnTF {\str_count:N #1} < {32} %or 31 or 33??
       { \exp_args:Ne \int_from_bin:n { #1 } }
       {
         \fp_eval:n { \int_step_function:nN {\str_count:N #1 }\__bitset_bin_factor:n  }
       }
  }     
 
 
\bitset_to_arabic:N \l_my_bitset \par

\bitset_set_true:Nn \l_my_bitset {45}

\bitset_to_arabic:N \l_my_bitset
\ExplSyntaxOff
\end{document}
~~~~

Reply:
That's a good idea, but the implementation should better be something like

    \cs_set:Npn \bitset_to_int:N #1 % this is really a much better name than arabic
      {
        \int_compare:nNnTF { \str_count:N #1 } < {32}
          { \exp_args:Ne \int_from_bin:n {#1} } % why "Ne" and not just "No" ??
          {
            \exp_after:wN \@@_to_int:nN \exp_after:wN 0
            #1 \q_recursion_tail \q_recursion_stop
          }
      }
    \cs_new:Npn \@@_to_int:nN #1#2
      {
        \quark_if_recursion_tail_stop_do:Nn #2 {#1}
        \exp_args:Nf \@@_to_int:nN { \fp_eval:n { #1 * 2 + #2 } }
      }

Not tested but hopefully clear.

Reply:
works fine (and the Ne is either a remainder from some older code or a typo).  But I don't agree really about the int. For me `int` has no intrinsic base. We have already `\bitset_to_bin`, and perhaps there will be a `\bitset_to_hex`.  `\bitset_to_int`  doesn't fit well (and we don't have an `\int_to_int:n`) .